modules/sr/robot3/arduino_devices.py

```diff
@@ -0,0 +1,250 @@
+from __future__ import annotations
+
+import abc
+import enum
+import random
+import logging
+from typing import Protocol
+
+from controller import (
+    LED,
+    Robot,
+    TouchSensor,
+    DistanceSensor as WebotsDistanceSensor,
+)
+from sr.robot3.utils import map_to_range, get_robot_device
+from sr.robot3.randomizer import add_jitter
+from sr.robot3.output_frequency_limiter import OutputFrequencyLimiter
+
+LOGGER = logging.getLogger(__name__)
+
+
+class GPIOPinMode(str, enum.Enum):
+    """The possible modes for a GPIO pin."""
+    INPUT = 'INPUT'
+    INPUT_PULLUP = 'INPUT_PULLUP'
+    OUTPUT = 'OUTPUT'
+
+
+DIGITAL_READ_MODES = {GPIOPinMode.INPUT, GPIOPinMode.INPUT_PULLUP, GPIOPinMode.OUTPUT}
+DIGITAL_WRITE_MODES = {GPIOPinMode.OUTPUT}
+ANALOG_READ_MODES = {GPIOPinMode.INPUT}
+
+
+class Device(Protocol):
+    def analogue_read(self) -> float:
```

Comment:
Should this be 'analog' for consistency?

Reply:
a87a0260048aeb62b593d22414f6da31e15d06c0